File: app/src/main/java/pl/szczodrzynski/edziennik/ui/dialogs/event/EventListAdapter.java
package pl.szczodrzynski.edziennik.ui.dialogs.event;

import android.content.Context;
import android.graphics.PorterDuff;
import android.graphics.PorterDuffColorFilter;
import androidx.annotation.NonNull;
import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.recyclerview.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.mikepenz.iconics.view.IconicsImageView;
import com.mikepenz.iconics.view.IconicsTextView;

import java.util.List;

import pl.szczodrzynski.edziennik.App;
import pl.szczodrzynski.edziennik.R;
import pl.szczodrzynski.edziennik.datamodels.Event;
import pl.szczodrzynski.edziennik.datamodels.EventFull;
import pl.szczodrzynski.edziennik.utils.models.Date;
import pl.szczodrzynski.edziennik.utils.Utils;

import static pl.szczodrzynski.edziennik.datamodels.Event.TYPE_HOMEWORK;
import static pl.szczodrzynski.edziennik.utils.Utils.bs;
import static pl.szczodrzynski.edziennik.utils.Utils.d;

public class EventListAdapter extends RecyclerView.Adapter<EventListAdapter.ViewHolder> {
    private static final String TAG = "EventListAdapter";
    private Context context;
    private List<EventFull> examList;
    private EventListDialog parentDialog;

    //getting the context and product list with constructor
    public EventListAdapter(Context mCtx, List<EventFull> examList, EventListDialog parentDialog) {
        this.context = mCtx;
        this.examList = examList;
        this.parentDialog = parentDialog;
    }

    @NonNull
    @Override
    public EventListAdapter.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        //inflating and returning our view holder
        LayoutInflater inflater = LayoutInflater.from(context);
        View view = inflater.inflate(R.layout.row_dialog_event_list_item, parent, false);
        return new EventListAdapter.ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull EventListAdapter.ViewHolder holder, int position) {
        App app = (App) context.getApplicationContext();

        EventFull event = examList.get(position);

        if (event.type == TYPE_HOMEWORK) {
            holder.eventListItemRoot.getBackground().setColorFilter(new PorterDuffColorFilter(0xffffffff, PorterDuff.Mode.CLEAR));
        }
        else {
            holder.eventListItemRoot.getBackground().setColorFilter(new PorterDuffColorFilter(event.getColor(), PorterDuff.Mode.MULTIPLY));
        }

        holder.eventListItemStartTime.setText(event.startTime == null ? app.getString(R.string.event_all_day) : event.startTime.getStringHM());
        //holder.examListItemEndTime.setText(event.endTime.getStringHM());
        holder.eventListItemTeamName.setText(Utils.bs(event.teamName));
        holder.eventListItemTeacherName.setText(bs(null, event.teacherFullName, "\n") + bs(event.subjectLongName));
        holder.eventListItemAddedDate.setText(Date.fromMillis(event.addedDate).getFormattedStringShort());
        holder.eventListItemType.setText(event.typeName);
        holder.eventListItemTopic.setText(event.topic);
        holder.eventListItemEdit.setVisibility((event.addedManually ? View.VISIBLE : View.GONE));
        holder.eventListItemHomework.setVisibility((event.type == Event.TYPE_HOMEWORK ? View.VISIBLE : View.GONE));
        holder.eventListItemEdit.setOnClickListener(v -> {
            if (parentDialog != null) {
                parentDialog.callDismissListener = false;
                parentDialog.dialog.dismiss();
            }
            d(TAG, "Event "+event);
            if (event.type == Event.TYPE_HOMEWORK) {
                new EventManualDialog(context, event.profileId)
                        .withDismissListener(parentDialog::performDismiss)
                        .show(app, event, null, null, EventManualDialog.DIALOG_HOMEWORK);
            }
            else {
                new EventManualDialog(context, event.profileId)
                        .withDismissListener(parentDialog::performDismiss)
                        .show(app, event, null, null, EventManualDialog.DIALOG_EVENT);
            }
        });

        if (event.sharedBy == null) {
            holder.eventListItemSharedBy.setVisibility(View.GONE);
        }
        else if (event.sharedByName != null) {
            holder.eventListItemSharedBy.setText(app.getString(R.string.event_shared_by_format, (event.sharedBy.equals("self") ? app.getString(R.string.event_shared_by_self) : event.sharedByName)));
        }
    }

    @Override
    public int getItemCount() {
        return examList.size();
    }

    class ViewHolder extends RecyclerView.ViewHolder {

        ConstraintLayout eventListItemRoot;
        TextView eventListItemStartTime;
        TextView eventListItemTeacherName;
        TextView eventListItemType;
        TextView eventListItemTopic;
        TextView eventListItemAddedDate;
        TextView eventListItemTeamName;
        IconicsImageView eventListItemEdit;
        IconicsImageView eventListItemHomework;
        IconicsTextView eventListItemSharedBy;


        ViewHolder(View itemView) {
            super(itemView);
            eventListItemRoot = itemView.findViewById(R.id.eventListItemRoot);
            eventListItemStartTime = itemView.findViewById(R.id.eventListItemStartTime);
            eventListItemTeacherName = itemView.findViewById(R.id.eventListItemTeacherName);
            eventListItemType = itemView.findViewById(R.id.eventListItemType);
            eventListItemTopic = itemView.findViewById(R.id.eventListItemTopic);
            eventListItemAddedDate = itemView.findViewById(R.id.eventListItemAddedDate);
            eventListItemTeamName = itemView.findViewById(R.id.eventListItemTeamName);
            eventListItemEdit = itemView.findViewById(R.id.eventListItemEdit);
            eventListItemHomework = itemView.findViewById(R.id.eventListItemHomework);
            eventListItemSharedBy = itemView.findViewById(R.id.eventListItemSharedBy);
        }
    }
}

File: app/src/main/java/pl/szczodrzynski/edziennik/ui/dialogs/event/EventListAdapter.kt
/*
 * Copyright (c) Kacper Ziubryniewicz 2019-11-30
 */

package pl.szczodrzynski.edziennik.ui.dialogs.event

import android.content.Context
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.core.view.isVisible
import androidx.recyclerview.widget.RecyclerView
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import pl.szczodrzynski.edziennik.*
import pl.szczodrzynski.edziennik.data.db.full.EventFull
import pl.szczodrzynski.edziennik.databinding.EventListItemBinding
import pl.szczodrzynski.edziennik.utils.models.Date
import pl.szczodrzynski.edziennik.utils.models.Week
import kotlin.coroutines.CoroutineContext

class EventListAdapter(
        val context: Context,
        val simpleMode: Boolean = false,
        val showWeekDay: Boolean = false,
        val showDate: Boolean = false,
        val showType: Boolean = true,
        val showTime: Boolean = true,
        val showSubject: Boolean = true,
        val markAsSeen: Boolean = true,
        val onItemClick: ((event: EventFull) -> Unit)? = null,
        val onEventEditClick: ((event: EventFull) -> Unit)? = null
) : RecyclerView.Adapter<EventListAdapter.ViewHolder>(), CoroutineScope {

    private val app = context.applicationContext as App
    private val manager
        get() = app.eventManager

    private val job = Job()
    override val coroutineContext: CoroutineContext
        get() = job + Dispatchers.Main

    var items = listOf<EventFull>()

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val inflater = LayoutInflater.from(parent.context)
        val view = EventListItemBinding.inflate(inflater, parent, false)
        return ViewHolder(view)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val event = items[position]
        val b = holder.b
        val manager = app.eventManager

        b.root.onClick {
            onItemClick?.invoke(event)
            if (!event.seen) {
                manager.markAsSeen(event)
            }
            if (event.showAsUnseen == true) {
                event.showAsUnseen = false
                notifyItemChanged(event)
            }
        }

        val bullet = " • "

        b.simpleMode = simpleMode

        manager.setEventTopic(b.topic, event, showType = false)
        b.topic.maxLines = if (simpleMode) 2 else 3

        b.details.text = mutableListOf<CharSequence?>(
                if (showWeekDay) Week.getFullDayName(event.date.weekDay) else null,
                if (showDate) event.date.getRelativeString(context, 7) ?: event.date.formattedStringShort else null,
                if (showType) event.typeName else null,
                if (showTime) event.time?.stringHM ?: app.getString(R.string.event_all_day) else null,
                if (showSubject) event.subjectLongName else null
        ).concat(bullet)

        b.addedBy.setText(
                when (event.sharedBy) {
                    null -> when {
                        event.addedManually -> R.string.event_list_added_by_self_format
                        event.teacherName == null -> R.string.event_list_added_by_unknown_format
                        else -> R.string.event_list_added_by_format
                    }
                    "self" -> R.string.event_list_shared_by_self_format
                    else -> R.string.event_list_shared_by_format
                },
                Date.fromMillis(event.addedDate).formattedString,
                event.sharedByName ?: event.teacherName ?: "",
                event.teamName?.let { bullet+it } ?: ""
        )

        b.typeColor.background?.setTintColor(event.eventColor)
        b.typeColor.isVisible = showType

        b.editButton.isVisible = !simpleMode && event.addedManually && !event.isDone
        b.editButton.onClick {
            onEventEditClick?.invoke(event)
        }
        b.editButton.attachToastHint(R.string.hint_edit_event)

        if (event.showAsUnseen == null)
            event.showAsUnseen = !event.seen

        b.unread.isVisible = event.showAsUnseen == true
        if (markAsSeen && !event.seen) {
            manager.markAsSeen(event)
        }
    }

    private fun notifyItemChanged(model: Any) {
        startCoroutineTimer(1000L, 0L) {
            val index = items.indexOf(model)
            if (index != -1)
                notifyItemChanged(index)
        }
    }

    override fun getItemCount() = items.size

    class ViewHolder(val b: EventListItemBinding) : RecyclerView.ViewHolder(b.root)
}
